TrackletAlgorithm/MatchProcessor.h

```diff
@@ -0,0 +1,903 @@
+                   ap_uint<NBits_Entries> nentries[nMEM],
+                   const int i,
+                   const MemType* const mem)
+  {    
+#pragma HLS inline  
+    ap_uint<kNBits_MemAddr+1> num = mem->getEntries(bx);
+    nentries[i] = num;
+    if (num > 0) mem_hasdata.set(i);
+  }
+  
+  // recursive case
+  template<int nMEM, int NBits_Entries, class MemType, class... Args>
+  inline void init(BXType bx, ap_uint<nMEM>& mem_hasdata,
+                   ap_uint<NBits_Entries> nentries[nMEM],
+                   const int i,
+                   const MemType* const mem, Args... args)
+  {
+#pragma HLS inline 
+    ap_uint<kNBits_MemAddr+1> num = mem->getEntries(bx);
+    nentries[i] = num;
+    if (num > 0) mem_hasdata.set(i);
+
+    if (i+1 < nMEM) init(bx, mem_hasdata, nentries, i+1, args...);
+  }
+  
+  //////////////////////////////
+  // Priority encoder based input memory reading logic
+  template<class DataType, class MemType>
+  void read_inmem(DataType& data, BXType bx, ap_uint<5> read_imem,
+                  ap_uint<kNBits_MemAddr>& read_addr,
+                  const int i, const MemType* const inmem)
+  {
+#pragma HLS inline
+    
+    if (read_imem == i) {
+      data = inmem->read_mem(bx, read_addr);
+    }
+  }
+
+  template<class DataType, class MemType, class... Args>
+  void read_inmem(DataType& data, BXType bx, ap_uint<5> read_imem,
+                  ap_uint<kNBits_MemAddr>& read_addr,
+                  const int i,
+                  const MemType* const inmem, Args... args)
+  {
+    if (read_imem == i) {
+      data = inmem->read_mem(bx, read_addr);
+    }
+    read_inmem(data, bx, read_imem, read_addr, i+1, args...);
+  }
+
+  template<class DataType, class MemType, int nMEM, int NBits_Entries>
+  bool read_input_mems(BXType bx,
+                       ap_uint<nMEM>& mem_hasdata,
+                       ap_uint<NBits_Entries> nentries[nMEM],
+                       ap_uint<kNBits_MemAddr>& read_addr,
+                       // memory pointers
+                       const MemType* const mem0, const MemType* const mem1,
+                       const MemType* const mem2, const MemType* const mem3,
+                       const MemType* const mem4, const MemType* const mem5,
+                       const MemType* const mem6, const MemType* const mem7,
+                       const MemType* const mem8, const MemType* const mem9,
+                       const MemType* const mem10, const MemType* const mem11,
+                       const MemType* const mem12, const MemType* const mem13,
+                       const MemType* const mem14, const MemType* const mem15,
+                       const MemType* const mem16, const MemType* const mem17,
+                       const MemType* const mem18, const MemType* const mem19,
+                       const MemType* const mem20, const MemType* const mem21,
+                       const MemType* const mem22, const MemType* const mem23,
```

Comment:
Don't know how much work it would be, but one day I guess these memories could be put in an array instead?

Reply:
Done.

Reply:
Done.